pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+// can be used to provide a more helpful error message
+func (v *PulumiInputValidator) getArrayTypeSpec(
+	spec *pschema.TypeSpec,
+	propertyValue resource.PropertyValue,
+) (
+	typeSpec *pschema.TypeSpec,
+	requiredProps []string,
+) {
+	if spec.Items != nil {
+		typeSpec = spec.Items
+	}
+	if len(spec.OneOf) > 0 {
+		typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+	}
+
+	return typeSpec, requiredProps
+}
+
+// getObjectTypeSpec gets a type spec for an object type. Objects can have their
+// type defined in a couple different ways. This will check the different ways
+// and return the correct type spec.
+func (v *PulumiInputValidator) getObjectTypeSpec(
+	spec *pschema.TypeSpec,
+	propertyValue resource.PropertyValue,
+) (
+	complexSpec *pschema.ComplexTypeSpec,
+	typeSpec *pschema.TypeSpec,
+	requiredProps []string,
+) {
+
+	if spec.AdditionalProperties != nil {
+		if spec.AdditionalProperties.Ref != "" {
+			complexSpec = v.getType(spec.AdditionalProperties.Ref)
+			// if we don't find the reference then just continue instead of failing
+			if complexSpec != nil {
+				requiredProps = complexSpec.Required
+			}
+		} else if len(spec.AdditionalProperties.OneOf) > 0 {
+			typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		} else {
+			typeSpec = spec.AdditionalProperties
+		}
+	} else if spec.Ref != "" {
+		complexSpec = v.getType(spec.Ref)
+		if complexSpec != nil {
+			requiredProps = complexSpec.Required
+		}
+	} else if len(spec.OneOf) > 0 {
+		typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		// If the OneOf that we found is an object with a `Type` then return that sub type
+		// Otherwise we should just return the OneOf type that we found
+		if typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type != "" {
+			typeSpec = typeSpec.AdditionalProperties
+		}
+	}
+	return complexSpec, typeSpec, requiredProps
+}
+
+// validateTypeSpec is the main function for validating a PropertyValue against the pulumi schema.
+// It is a recursive function that will validate nested types and arrays. It will return a list of
+// type failures if any are found
+func (v *PulumiInputValidator) validateTypeSpec(
+	propertyKey string,
+	propertyValue resource.PropertyValue,
+	typeSpec pschema.TypeSpec,
+	pathBuilder pathBuilder,
+	requiredProps []string,
+) *[]TypeFailure {
+	// for now don't validate discriminators
+	if typeSpec.Discriminator != nil {
+		return nil
+	}
+	// don't type check
+	// - resource references
+	// - assets
+	// - archives
+	if propertyValue.IsResourceReference() || propertyValue.IsAsset() || propertyValue.IsArchive() {
+		return nil
+	}
+
+	// handle non-required properties that are null
+	if propertyValue.TypeString() == "null" {
+		if slices.Contains(requiredProps, propertyKey) {
+			return &[]TypeFailure{
+				{
+					ResourcePath: pathBuilder.toPath(),
+					Reason:       fmt.Sprintf("property %s is required", propertyKey),
```

Comment:
It seems that this check is redundant with `!slices.Contains(stableKeys, resource.PropertyKey(requiredProp))` done when checking objects? Or am I missing something?

Reply:
The property may be a top level property that is not an object property.

Reply:
Ah yes that's right.. That's good then. There's another weird way to handle it is wrapping the top-level property-set schema for a resource in an object although it isn't, but it can be confusing. Sounds good then.